src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1063 @@
+                if(mat.num_faces == 0) continue;
+
+                MeshChunk &chunk = _state->meshChunks[chunkOffset];
+                chunk.meshId = mesh->typed_id;
+                chunk.meshMaterialIndex = UnsignedInt(i);
+                ++chunkOffset;
+            }
+        }
+    }
+
+    /* Count the final number of nodes in the scene, we may remove some (root)
+       or add (synthetic geometry transform nodes) */
+    {
+        const bool preserveRootNode = configuration().value<bool>("preserveRootNode");
+        const bool geometricTransformNodes = configuration().value<bool>("geometricTransformNodes");
+
+        _state->nodeIdOffset = 0;
+        _state->originalNodeCount = UnsignedInt(scene->nodes.count);
+
+        if(!preserveRootNode) {
+            --_state->originalNodeCount;
+            ++_state->nodeIdOffset;
+        }
+
+        _state->nodeCountWithSynthetic = _state->originalNodeCount;
+
+        /* Reserve space for nodes if we want to create dummy nodes for geometric
+           transforms */
+        if(geometricTransformNodes) {
+            for(ufbx_node *node : scene->nodes) {
+                if(node->has_geometry_transform) {
+                    ++_state->nodeCountWithSynthetic;
+                }
+            }
+        }
+    }
+}
+
+Int UfbxImporter::doDefaultScene() const { return 0; }
+
+UnsignedInt UfbxImporter::doSceneCount() const { return 1; }
+
+Containers::Optional<SceneData> UfbxImporter::doScene(UnsignedInt) {
+    ufbx_scene *scene = _state->scene.get();
+
+    const bool preserveRootNode = configuration().value<bool>("preserveRootNode");
+    const bool geometricTransformNodes = configuration().value<bool>("geometricTransformNodes");
+
+    UnsignedInt meshCount = 0;
+    UnsignedInt skinCount = 0;
+    UnsignedInt nodeCount = _state->nodeCountWithSynthetic;
+    UnsignedInt cameraCount = 0;
+    UnsignedInt lightCount = 0;
+
+    /* ufbx meshes can contain per-face materials so we need to separate them
+       into pieces containing a single material for SceneData. */
+    for(std::size_t i = 0; i < scene->meshes.count; ++i) {
+        ufbx_mesh *mesh = scene->meshes[i];
+
+        UnsignedInt instanceCount = UnsignedInt(mesh->instances.count);
+        for(const ufbx_mesh_material &mat : mesh->materials) {
+            if(mat.num_faces == 0) continue;
+            meshCount += instanceCount;
+            if(mesh->skin_deformers.count > 0)
+                skinCount += instanceCount;
+        }
+    }
+
+    /* Collect instanced camera/light counts */
+    for(ufbx_light *light : scene->lights)
+        lightCount += UnsignedInt(light->instances.count);
+    for(ufbx_camera *camera : scene->cameras)
+        cameraCount += UnsignedInt(camera->instances.count);
+
+    /* Allocate the output array. */
+    Containers::ArrayView<UnsignedInt> nodeObjects;
+    Containers::ArrayView<Int> parents;
+    Containers::ArrayView<Matrix4x3d> transformations;
```

Comment:
One thing I realized just now -- if you provide TRS components for all nodes (which seems like you do), you don't need to supply the matrix as well -- the `SceneData` will calculate it on the fly if needed. Supplying the matrix would be only needed if you have TRS components for just a subset of the nodes, with the rest having just a matrix.

Reply:
Ah that's good to know, ufbx internally just converts the transform to a matrix for user convenience, so letting Magnum do it is a bit more memory efficient and should be identical!

```c
node->node_to_parent = ufbx_transform_to_matrix(&node->local_transform); 
```